ui/src/wizard/LocationListStep.ts

```diff
@@ -22,50 +19,108 @@ interface ILocationWizardContextInternal extends types.ILocationWizardContext {
-        name = generalizeLocationName(name);
-        wizardContext.location = locations.find(l => {
-            return name === generalizeLocationName(l.name) || name === generalizeLocationName(l.displayName);
-        });
+        const [allLocationsTask] = this.getInternalVariables(wizardContext);
+        wizardContext._location = (await allLocationsTask).find(l => matchesLocation(l, name));
     }
 
-    public static async getLocations<T extends ILocationWizardContextInternal>(wizardContext: T): Promise<types.AzExtLocation[]> {
-        if (wizardContext._allLocationsTask === undefined) {
-            wizardContext._allLocationsTask = getAllLocations(wizardContext);
-        }
+    public static setLocationSubset<T extends ILocationWizardContextInternal>(wizardContext: T, task: Promise<string[]>, provider: string): void {
+        const [, providerLocationsMap] = this.getInternalVariables(wizardContext);
+        providerLocationsMap.set(provider.toLowerCase(), task);
+    }
 
-        const allLocations = await wizardContext._allLocationsTask;
-        if (wizardContext.locationsTask === undefined) {
-            return allLocations;
-        } else {
-            const locationsSubset: { name?: string }[] = await wizardContext.locationsTask;
-            return allLocations.filter(l1 => locationsSubset.find(l2 => generalizeLocationName(l1.name) === generalizeLocationName(l2.name)));
+    public static hasLocation<T extends ILocationWizardContextInternal>(wizardContext: T): boolean {
+        return !!wizardContext._location;
+    }
+
+    public static async getLocation<T extends ILocationWizardContextInternal>(wizardContext: T, provider?: string): Promise<types.AzExtLocation> {
+        const location = nonNullProp(wizardContext, '_location');
+        if (provider) {
+            const [allLocationsTask, providerLocationsMap] = this.getInternalVariables(wizardContext);
+            const providerLocations = await providerLocationsMap.get(provider.toLowerCase());
+            if (providerLocations) {
+                function isSupportedByProvider(loc: types.AzExtLocation): boolean {
+                    return !!providerLocations?.find(name => matchesLocation(loc, name));
+                }
+                function useProviderName(loc: types.AzExtLocation): types.AzExtLocation {
+                    // Some providers prefer their version of the name over the standard one, so we'll create a shallow clone using theirs
+                    return { ...loc, name: nonNullValue(providerLocations?.find(name => matchesLocation(loc, name), 'providerName')) };
+                }
+
+                if (isSupportedByProvider(location)) {
+                    return useProviderName(location);
+                }
+
+                if (location.name.toLowerCase().endsWith('stage')) {
+                    const nonStageName = location.name.replace(/stage/i, '');
```

Comment:
I don't think we have any actual location name with stage in it. But this could potentially remove the 'stage' which is not at the end.

Reply:
I think I'm fine with that. For example, may there's "North Central US (Stage) 2" or "(Stage) North Central US"

In any case, the important thing to note is that I still check `isSupportedByProvider` before actually using this `nonStageName`. I don't think there's much possible harm since I'm doing that. Although I suppose I could add a warning in the output channel